[GEP-33] Machine Image Capabilities (#11048)
gardener/gardener: docs/proposals/33-machine-image-capabilities.md

    network: ["accelerated", "standard"] # <-- replaces acceleratedNetworking

  machineImages:
    - name: gardenlinux
      updateStrategy: minor
      versions:
        # - architectures: [arm64, amd64] # not required anymore
        #   version: 1592.2.0-gen2

        - architectures: [amd64, arm64]
          capabilities:
            - architecture: [arm64]
              hypervisorType: ["gen2"]
              network: ["accelerated", "standard"] # <-- not required as its the default
            - architecture: [amd64]
              hypervisorType: ["gen2"]
            - architecture: [amd64]
              hypervisorType: ["gen1"]
          classification: supported
          version: 1592.2.0

  machineTypes:
    - architecture: amd64
      cpu: "896"
      gpu: "0"
      memory: 12Ti
      name: Standard_S896om
      usable: true
      capabilities:
        - hypervisorType: ["gen2"] # <-- hypervisorType is overwritten

  providerConfig:
    machineImages:
      - name: gardenlinux
        versions:
          - architecture: arm64 # <-- architecture is overwritten as only amd64 is supported
            capabilities:
              - hypervisorType: ["gen2"]

>               - network: ["accelerated", "standard"] # <-- not required as its the default

Reviewer: Maybe one should be explicit here and use at least an alias to the default capabilities instead of implicitly assuming (default) capabilities  

(also refers to [line 240](https://github.com/gardener/gardener/pull/11048/files#diff-08a10b2ab42511d2f0727d1e6cbb8f1dd39014b96ad6e57774dc86d6453c5746R240))

Author: I think it is reasonable to work with defaults here. My main arguments are that it improves readability and reduces the cloud profile size significantly.

---
title: Machine Image Capabilities
gep-number: 33
creation-date: 2024-12-16
status: implementable
authors:
  - "@MrBatschner"
  - "@Roncossek"
reviewers:
  - "@rfranzke"

---

# GEP-33: Machine Image Capabilities

## Table of Contents

- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
- [Alternatives](#alternatives)

## Summary

This GEP proposes the introduction of capabilities for machineType and machineImage in the Gardener CloudProfile to ensure only shoots with a matching set of capabilities on machine type and image will be admitted by Gardener.
This will prevent the creation of incompatible worker pools on shoot creation and maintenance operations.

> It will also provide an mechanism to filter out incompatible images for a machine type in the Gardener Dashboard.

Reviewer 2: ```suggestion
It will also provide a mechanism to filter out incompatible images for a machine type in the Gardener Dashboard.
```

While `gen1` machine types boot with legacy BIOS support, `gen2` machine types make use of UEFI.
While most machine types actually support both, `gen1` and `gen2`, there are certain machine types which support only one.
Machine images in Azure also have the generation in their metadata, however, while a machine type can be `gen1` **and** `gen2`, a machine image can only be `gen1` **or** `gen2`.
For machine types, that support either generation, it depends on the selected machine image wether the machine will be booted with legacy BIOS or with UEFI.
For those machines however, that only support a single generation, providing a mismatching machine image in the `create_or_update` API call will result in an error.

The information given on the Gardener Dashboard provide no information about that incompatibility.
The user has to know about that and select the proper machine image manually AND disable Operating System maintenance as the machine image might be updated to a version that is incompatible with the machine type.

![](assets/gep-33-screenshot-dashboard.png)

### Context: Cloud Profile

> For each infrastructure provider Gardener operators must provide CloudProfile.

Reviewer 2: ```suggestion
For each infrastructure provider Gardener operators must provide a CloudProfile.
```

As images with pre-release tag will not be regarded during maintenance operations, the user has to disable Operating System maintenance for the shoot.

2. In the cloud profile only one OS image version can be in state `classification: supported`.
When another capability set for that version like Azures `generation` must to be supported, the `classification` key is omitted by Gardener Operators.
This will display the version as supported in the Gardener Dashboard.
This is not the intended use of the classification key.

3. Currently extension providers define their own capabilities that are processed after shoot admission, like `acceleratedNetworking` in the [gardener-extension-provider-azure](https://github.com/gardener/gardener-extension-provider-azure/blob/28c977612898ed40e9d179052633fee0b9600d3e/pkg/apis/azure/types_cloudprofile.go#L78).
This can lead to a situation where the machine supports accelerated networking but the machine image does not.

> In this case it does not result in an error but to performance loss.

Reviewer 2: ```suggestion
In this case it does not result in an error but in performance loss.
```

    network: ["accelerated", "standard"] # <-- replaces acceleratedNetworking

  machineImages:
    - name: gardenlinux
      updateStrategy: minor
      versions:
        # - architectures: [arm64, amd64] # not required anymore
        #   version: 1592.2.0-gen2

        - architectures: [amd64, arm64] 
          capabilities:
            - architecture: [arm64] # <-- architecture must be added to the capabilities to ensure compatibility
              hypervisorType: ["gen2"]
              network: ["accelerated", "standard"] # <-- not required as its the default
            - architecture: [amd64]
              hypervisorType: ["gen2"]
            - architecture: [amd64]
              hypervisorType: ["gen1"]
          classification: supported
          version: 1592.2.0

  machineTypes:
    - architecture: amd64
      cpu: "896"
      gpu: "0"
      memory: 12Ti
      name: Standard_S896om
      usable: true
      capabilities:
        - hypervisorType: ["gen2"] # <-- hypervisorType is overwritten

  providerConfig:
    machineImages:
      - name: gardenlinux
        versions:
          - architecture: arm64 # <-- architecture is overwritten as only amd64 is supported

>             capabilities:
>               - hypervisorType: ["gen2"]
>               - network: ["accelerated", "standard"] # <-- not required as its the default

Reviewer 2: Why are the `capabilities` individual arrays here? Wouldn't it rather be a map itself as it's already defined per architecture?

```suggestion
            capabilities:
              hypervisorType: ["gen2"]
              network: ["accelerated", "standard"] # <-- not required as its the default
```

Author: @LucaBernstein thank you for your contribution! 🙏
I agree a map makes more sense. I will adjust the GEP accordingly.

          - architecture: amd64
            capabilities:
              - hypervisorType: ["gen2"]
            communityGalleryImageID: /CommunityGalleries/xzy/Images/gardenlinux-nvme-gen2/Versions/1592.2.0
            version: 1592.2.0
          - architecture: amd64
            capabilities:
              - hypervisorType: ["gen1"]
            communityGalleryImageID: /CommunityGalleries/xzy/Images/gardenlinux-nvme/Versions/1592.2.0
            version: 1592.2.0

    # machineTypes: # <--- not required anymore
    #   - acceleratedNetworking: true
    #     name: Standard_S896om
```

The algorithm to determine if an image is valid for a machine type is given as follows: 
For every capability the union of values that is supported by the machine type and the machine image must not be empty.
In other words, the machine image must support at least one value of each capability of the machine type.

```js
for capabilityName, machineCapabilities in machineType.capabilities:
  if union(imageVersion.capabilities[capabilityName], machineCapabilities) is empty:
    return false
      
return true
```

### Migration Strategy

> Capabilities are **OPT-IN** and not all infrastructure providers or Gardener installation might use them.

Reviewer 2: ```suggestion
Capabilities are **OPT-IN** and not all infrastructure providers or Gardener installations might use them.
```

---
title: Machine Image Capabilities
gep-number: 33
creation-date: 2024-12-16
status: implementable
authors:
  - "@MrBatschner"
  - "@Roncossek"
reviewers:

>   - "@rfranzke"

Reviewer 3: ```suggestion
  - "@rfranzke"
  - "@LucaBernstein"
  - "@timuthy"
```

---

# GEP-33: Machine Image Capabilities

## Table of Contents

- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
- [Alternatives](#alternatives)

## Summary

This GEP proposes the introduction of capabilities for machineType and machineImage in the Gardener CloudProfile to ensure only shoots with a matching set of capabilities on machine type and image will be admitted by Gardener.
This will prevent the creation of incompatible worker pools on shoot creation and maintenance operations.
It will also provide a mechanism to filter out incompatible images for a machine type in the Gardener Dashboard.

> ❗️❗️❗️One decisions in this GEP is to deprecate the current architecture field in the CloudProfile. Otherwise the information will be duplicated within the CloudProfile as the architecture is mandatory to be included into capability combinations❗️❗️❗️

Reviewer 3: ```suggestion
> [!Note]
> One decision in this GEP is to deprecate the current architecture fields in the `CloudProfile`. Otherwise the information will be duplicated as the architecture is mandatory to be included into capability combinations.
```

2. Operating System Maintenance Operations

When performing maintenance operations on a cluster, Gardener has to know which machine image can be used for a given machine type.
This is important to ensure that the maintenance operation is performed correctly and does not break the worker nodes.

3. Gardener Dashboard Shoot Creation

During shoot creation via the Gardener Dashboard, the user has to select a machine type and a machine image.
Currently also incompatible combinations can be selected.

### Use Case Example

One such example is on Azure: machine types have a `generation` [attribute](https://learn.microsoft.com/en-us/azure/virtual-machines/generation-2) which - at the time of writing - supports two values `gen1` and `gen2`.
While `gen1` machine types boot with legacy BIOS support, `gen2` machine types make use of UEFI.
While most machine types actually support both, `gen1` and `gen2`, there are certain machine types which support only one.
Machine images in Azure also have the generation in their metadata, however, while a machine type can be `gen1` **and** `gen2`, a machine image can only be `gen1` **or** `gen2`.

> For machine types, that support either generation, it depends on the selected machine image wether the machine will be booted with legacy BIOS or with UEFI.

Reviewer 2: ```suggestion
For machine types, that support either generation, it depends on the selected machine image whether the machine will be booted with legacy BIOS or with UEFI.
```

For those machines however, that only support a single generation, providing a mismatching machine image in the `create_or_update` API call will result in an error.

The information given on the Gardener Dashboard provide no information about that incompatibility.
The user has to know about that and select the proper machine image manually AND disable Operating System maintenance as the machine image might be updated to a version that is incompatible with the machine type.

![](assets/gep-33-screenshot-dashboard.png)

> ### Context: Cloud Profile

Reviewer 3: ```suggestion
### Context: `CloudProfile`
```

    machineImages:
      - name: gardenlinux
        versions:
          - acceleratedNetworking: true
            architecture: arm64
            communityGalleryImageID: /CommunityGalleries/xzy/Images/gardenlinux-nvme-arm64-gen2/Versions/1592.2.0
            version: 1592.2.0-gen2 # first entry of 1592.2.0 with gen2 AND arm64
          - acceleratedNetworking: true
            architecture: amd64
            communityGalleryImageID: /CommunityGalleries/xzy/Images/gardenlinux-nvme-gen2/Versions/1592.2.0
            version: 1592.2.0-gen2 # second entry of 1592.2.0 with gen2 AND amd64
          - acceleratedNetworking: true
            architecture: amd64
            communityGalleryImageID: /CommunityGalleries/xzy/Images/gardenlinux-nvme/Versions/1592.2.0
            version: 1592.2.0 # third entry of 1592.2.0 with gen1 AND amd64

    machineTypes: # provider specific capabilities
      - acceleratedNetworking: true
        name: Standard_S896om
```

</details></br>

While a machineType in `spec.machineType` always refers to exactly one resource in the cloud provider, the `spec.machineImages` section is a bit confusing.

> A machineImage defines an operating system, like gardenLinux or SUSE.

Reviewer 3: ```suggestion
A machineImage defines an operating system, like GardenLinux or SUSE.
```

- Minimize the additional data size introduced to cloud profiles to prevent bloating.
- Obsolete existing workarounds described in the "Current Workarounds" section.

### Non-Goals

- Introduce capability options that are not utilized by Gardener in the future.
- Introduce feature flag to control the usage of a capability, e.g. secure boot.

## Proposal

Introduce a top level capabilities map in the CloudProfile `spec.capabilities`.

Capabilities are very specific to the provider and the selected catalog offered by Gardener in the cloud profile.
To minimize complexity and data size in the cloud profile, the capabilities are defined as a map with string keys and string arrays as values.
The key is the capability name and the value is an array of possible values.

```go
type Spec struct {
    Capabilities map[string][]string `json:"capabilities"`
}
```

For each cloud profile the capabilities are defined in the `spec.capabilities` map.
The full set of possibilities for each capability is defined here.
As some capabilities can have multiple values at the same time an array of possible values is used instead of a single value.

> This structure defines the default set of capabilities.

Reviewer 3: ```suggestion
This structure defines the **default set** of capabilities.
```

> If no further information are provided each machine type and machine image will be assigned the default set of capabilities.

Reviewer 3: ```suggestion
If no further information is provided each machine type and machine image will be assigned the default set of capabilities.
```

> Thus being compatible with each other.

Reviewer 3: ```suggestion
```
Not quite sure what this sentence says in this context. Remove or rephrase?

```yaml
# CloudProfile Example
spec:
  capabilities:
    hypervisorType: ["gen2", "gen1"]
    network: ["accelerated", "standard"]
    storageAccess: ["NVMe", "SCSI"]
    secureBoot: ["secure", "none"]

>     bootMode: ["uefi-prefered", "uefi", "legacy-bios"]

Reviewer 2: ```suggestion
    bootMode: ["uefi-preferred", "uefi", "legacy-bios"]
```

    ...
```

> The array of a capability is ordered.
> The order defines the priority of the values in case of multiple supported images.
> The first value in the array is the most preferred value.
> The last value is the least preferred value.

Reviewer 3: ```suggestion
Please note the following characteristics:
- The array of a capability is ordered.
- The order defines the priority of the values in case of multiple supported images.
- The first value in the array is the most preferred value.
- The last value is the least preferred value.
```

Example: A machine supports hypervisor `gen2` AND `gen1` and an image version offers `gen1` OR `gen2`.
Then the image with `gen2` will be preferred.

Other edge cases will be deterministic solved as well to ensure a consistent behavior.

In addition to the default capabilities, the `spec.machineTypes` is extended with the `capabilities` structure described above.
The image versions in `spec.machineImages.versions` will be extended with `capabilityCombinations` - an array of capabilities structures, one entry for each image reference of an image version in the cloud provider.

The architecture is also added to the capabilityCombinations. This is required as the architecture is a capability itself. 

> ❗️❗️❗️NOTE: To avoid duplication and remove the possibility of inconsistencies the existing `architecture` field on `machineImages.versions` and `machineTypes` will be marked as deprecated. Architecture will become the only mandatory capability and will take precedence over the deprecated field if both are present.

Reviewer 3: ```suggestion
> [!NOTE]
> To avoid duplication and remove the possibility of inconsistencies the existing `architecture` field in `machineImages.versions` and `machineTypes` will be marked as deprecated. `architecture` will become the only mandatory capability and will take precedence over the deprecated field if both are present.
```

```yaml
# CloudProfile
spec:
  capabilities: # <-- Full list of possible capabilities used as default
    architecture: ["amd64", "arm64"]
    hypervisorType: ["gen2", "gen1"]
    network: ["accelerated", "standard"]

  machineImages:
    - name: gardenlinux
      updateStrategy: minor
      versions:
        # - architectures: [arm64, amd64] # not required anymore
        #   version: 1592.2.0-gen2

        - architectures: [amd64, arm64] # <-- marked as deprecated

>           capabilityCombinations:

Reviewer 3: ```suggestion
          capabilitySets:
```

Thanks @LucaBernstein 🚀

            - architecture: [arm64] # <-- architecture must be added to the capabilities to ensure compatibility
              hypervisorType: ["gen2"]
              network: ["accelerated", "standard"] # <-- not required as its the default
            - architecture: [amd64]
              hypervisorType: ["gen2"]
            - architecture: [amd64]
              hypervisorType: ["gen1"]
          classification: supported
          version: 1592.2.0

        - classification: supported
          version: 1592.1.0 # <-- if no capabilityCombinations are defined the default capabilities are assigned as only combination

  machineTypes:
    - architecture: amd64 # <-- marked as deprecated
      cpu: "896"
      gpu: "0"
      memory: 12Ti
      name: Standard_S896om
      usable: true
      capabilities:
        hypervisorType: ["gen2"] # <-- hypervisorType is overwritten
        architecture: ["amd64"] # <-- architecture is overwritten

    - cpu: "896" # <-- as no capabilities are defined the default capabilities are assigned
      gpu: "0"
      memory: 12Ti
      name: Standard_S896
      usable: true
```

> The changes to the gardener CloudProfile above are sufficient to select the correct image version selection. This enables:

Reviewer 3: ```suggestion
The changes to the gardener `CloudProfile` above are sufficient to select the correct image version selection. This enables:
```

- The shoot admission controller to check if the selected machine image is supported by the selected machine type. 
- The maintenance controller to performed automated upgrades correctly without breaking worker nodes.
- The Gardener Dashboard to filter out incompatible images for a machine type.

> The selection of the actual image reference is still done by the provider extension. Therefore the provider extension must be updated to include the capabilities in the cloud profile as well in `spec.providerConfig.machineImages.versions` also here the architecture field will be marked as deprecated.

Reviewer 3: Does it make sense to mention that the capability combinations are expected to be found here? Also provider extensions are supposed to validate each versions' `capabilities` has a matching configuration in `capabilityCombinations`.

Author: copy that :)

          - capabilities:
              architecture: [amd64]
              hypervisorType: ["gen2"]
            communityGalleryImageID: /CommunityGalleries/xzy/Images/gardenlinux-nvme-gen2/Versions/1592.2.0
            version: 1592.2.0
          - capabilities:
              architecture: [amd64]
              hypervisorType: ["gen1"]
            communityGalleryImageID: /CommunityGalleries/xzy/Images/gardenlinux-nvme/Versions/1592.2.0
            version: 1592.2.0
```

The algorithm to determine if an image is valid for a machine type is given as follows: 

For every capability the union of values that is supported by the machine type and the machine image must not be empty.
In other words, the machine image must support at least one value of each capability of the machine type.

```js
for capabilityName, machineCapabilities in machineType.capabilities:
  if union(imageVersion.capabilities[capabilityName], machineCapabilities) is empty:
    return false
      
return true
```

## Implications on NamespacedCloudProfiles. 

The following implications are to be considered in regards to `NamespacedCloudProfile`s:
 

> * `NamespacedCloudProfile`s won't have a global `capabilities` definition

Reviewer 2: ```suggestion
* `NamespacedCloudProfile`s won't have a global `capabilities` definition.
```

* For overridden machine image versions, no `capabilities` must be defined, as they are inherited from the machine image versions of the parent `CloudProfile`.

> * For custom machine image versions, the `capabilities` need to be defined in the `NamespacedCloudProfile` as they would be in the `CloudProfile` (as well as for the `providerConfig`).

Reviewer 2: ```suggestion
* For custom machine image versions, the `capabilities` need to be defined in the `NamespacedCloudProfile`, as they would be in the `CloudProfile` (as well as for the `providerConfig`).
```

> * This has also implications on later-added equal machine image versions to the parent `CloudProfile`: On conflicting definition, the machine image version of the `NamespacedCloudProfile` will be rendered in the resulting status (see [Fix `NamespacedCloudProfile` validation for {`MachineImages`,`MachineTypes`} added to parent `CloudProfile` later #11093](https://github.com/gardener/gardener/pull/11093) for reference).

Reviewer 2: Imho, the PR title can be omitted in order to improve the flow of reading.

```suggestion
* This has also implications on later-added equal machine image versions to the parent `CloudProfile`: On conflicting definition, the machine image version of the `NamespacedCloudProfile` will be rendered in the resulting status (see [this PR](https://github.com/gardener/gardener/pull/11093) for reference).
```

### Migration Strategy

Capabilities are **OPT-IN** and not all infrastructure providers or Gardener installations might use them.
All API changes in Gardener core are backwards compatible.
They can be implemented first including its filter and admission logic.
In a second step each provider extension can be updated to include the capabilities in the cloud profile.
In a third wave cloud profiles and the Gardener Dashboard can be updated to react to the capabilities.

> ❗️❗️❗️ when the architecture field is marked as deprecated the eventual removal in the  CloudProfile API will be a breaking change. Latest this point all cloud profiles must be updated to include the architecture in the capabilities. It will no longer be opt in.❗️❗️❗️

Reviewer 3: ```suggestion
> [!IMPORTANT]
> When the `architecture` field is marked as deprecated the eventual removal in the `CloudProfile` API will be a breaking change. Latest at this point all `CloudProfile`s must be updated to include the architecture in the capabilities. It will no longer be opt in.
```

type Spec struct {
    Capabilities map[string][]string `json:"capabilities"`
}
```

For each cloud profile the capabilities are defined in the `spec.capabilities` map.
The full set of possibilities for each capability is defined here.
As some capabilities can have multiple values at the same time an array of possible values is used instead of a single value.

This structure defines the **default set** of capabilities.
If no further information is provided each machine type and machine image will be assigned the default set of capabilities.

```yaml
# CloudProfile Example
spec:
  capabilities:
    hypervisorType: ["gen2", "gen1"]
    network: ["accelerated", "standard"]
    storageAccess: ["NVMe", "SCSI"]
    secureBoot: ["secure", "none"]
    bootMode: ["uefi-preferred", "uefi", "legacy-bios"]
    ...
```

Please note the following characteristics:
- The array of a capability is ordered.
- The order defines the priority of the values in case of multiple supported images.
- The first value in the array is the most preferred value.
- The last value is the least preferred value.

Example: A machine supports hypervisor `gen2` AND `gen1` and an image version offers `gen1` OR `gen2`.

> Then the image with `gen2` will be preferred.

Reviewer 4: Are the `capabilities` ordered too?
In the example it could be the case that a machine supports hypervisor `gen1` and `gen2`, network `accelerated` and `standard`.
`Image1` supports `gen2` and `standard`. `Image2` supports `gen1` and `accelerated`. 
Which image will be selected?

Author: Thanks for your suggestions. One could also make the `capabilitiesDefinition` ordered but this leads to an increased complexity in the api design and its validation. The optimal way would be an arrays of maps here, each holding one capability. This is not possible with the current protobuf parser and the "raw" `JSON` type must be used....

I think this case is a rather niche szenario. I plan to implement a deterministic tiebreaker in that case, maybe alphabetical order.

Other edge cases will be deterministic solved as well to ensure a consistent behavior.

In addition to the default capabilities, the `spec.machineTypes` is extended with the `capabilities` structure described above.

> The image versions in `spec.machineImages.versions` will be extended with `capabilityCombinations` - an array of capabilities structures, one entry for each image reference of an image version in the cloud provider.

Reviewer 2: ```suggestion
The image versions in `spec.machineImages.versions` will be extended with `capabilitySets` - an array of capabilities structures, one entry for each image reference of an image version in the cloud provider.
```

> The architecture is also added to the capabilityCombinations. This is required as the architecture is a capability itself. 

Reviewer 2: ```suggestion
The architecture is also added to the capabilitySets. This is required as the architecture is a capability itself. 
```

```yaml
# CloudProfile
spec:
  capabilities: # <-- Full list of possible capabilities used as default
    architecture: ["amd64", "arm64"]
    hypervisorType: ["gen2", "gen1"]
    network: ["accelerated", "standard"]

  machineImages:
    - name: gardenlinux
      updateStrategy: minor
      versions:
        # - architectures: [arm64, amd64] # not required anymore
        #   version: 1592.2.0-gen2

        - architectures: [amd64, arm64] # <-- marked as deprecated
          capabilitySets:
            - architecture: [arm64] # <-- architecture must be added to the capabilities to ensure compatibility
              hypervisorType: ["gen2"]
              network: ["accelerated", "standard"] # <-- not required as its the default
            - architecture: [amd64]
              hypervisorType: ["gen2"]
            - architecture: [amd64]
              hypervisorType: ["gen1"]
          classification: supported
          version: 1592.2.0

        - classification: supported

>           version: 1592.1.0 # <-- if no capabilityCombinations are defined the default capabilities are assigned as only combination

Reviewer 2: ```suggestion
          version: 1592.1.0 # <-- if no capabilitySets are defined the default capabilities are assigned as only combination
```

  machineTypes:
    - architecture: amd64 # <-- marked as deprecated
      cpu: "896"
      gpu: "0"
      memory: 12Ti
      name: Standard_S896om
      usable: true
      capabilities:
        hypervisorType: ["gen2"] # <-- hypervisorType is overwritten
        architecture: ["amd64"] # <-- architecture is overwritten

    - cpu: "896" # <-- as no capabilities are defined the default capabilities are assigned
      gpu: "0"
      memory: 12Ti
      name: Standard_S896
      usable: true
```

> The changes to the gardener `CloudProfile` above are sufficient to select the correct image version selection. This enables:

Reviewer 2: This sentence reads a bit bumpy. Maybe rephrasing with `support` (or `enhance`, or `aid`) makes the intended statement clearer? Open for other suggestions.

```suggestion
The changes to the gardener `CloudProfile` above are sufficient to support the correct image version selection. This enables:
```

One mayor goal is that the number of manual interventions on cluster maintenance should be minimized.
This means that the capabilities should be designed in a way that they can be used to automate the selection of machine images and machine types even when multiple images are suitable for a machine type.

For most capabilities there is a clear direction in terms of deprecation.
Bios is being deprecated and UEFI is the future.
Or Images that support NVMe also support SCSI, but not the other way around.

That indicates that capability values should have a preferred value or better a clear order of preference in case multiple values are supported.
This way the selection of the machine image can be automated and it is ensured that clusters are updated into the preferred direction so less legacy systems have to be maintained.

> Note: Some capabilities like secure boot might be disrupting but the activation of such features must be designed in a separate feature.
Capabilities just ensure that a machine with a capability is served an image that also supports it.

### Component changes

The Capability struct must be added to the following component:

- Gardener Core:

>   - MachineImageVersion\* in [gardener/pkg/apis/core

Reviewer 2: we also have one here

Author: see above

    /types_cloudprofile.go](https://github.com/gardener/gardener/blob/c0486a3187b5e2e0697f9e373c008c588689960a/pkg/apis/core/types_cloudprofile.go#L96)
  - MachineType in [gardener/pkg/apis/core/types_cloudprofile.go](https://github.com/gardener/gardener/blob/c0486a3187b5e2e0697f9e373c008c588689960a/pkg/apis/core/types_cloudprofile.go#L121)
  - Worker in [gardener/pkg/apis/core
    /types_shoot.go](https://github.com/gardener/gardener/blob/c0486a3187b5e2e0697f9e373c008c588689960a/pkg/apis/core/types_shoot.go#L1161)

> \* an array of Capability structs must be added

Reviewer 2: One more question: What is it about this escaped asterisk? Is this intentional? Should this be part of some enumeration?

Reviewer 2: Ah, I think I got it now: This should act as a reference for the addition. Suggestion: We use `(*)` instead, to make it more visually apparent?

Else it's almost indistinguishable from the other enumeration bullet points imho...

Author: I wanted to have a real asterisk here. To indicate that here the capabilitySets are added instead of the single capability struct.

### Goals

- Define a pattern to ensure only shoots with compatible machine types and images are admitted by Gardener.
- Provide a mechanism to filter out incompatible images for a machine type in the Gardener Dashboard.
- Provide a pattern and process to add new or remove deprecated capabilities in the future.
- Minimize the additional data size introduced to cloud profiles to prevent bloating.
- Obsolete existing workarounds described in the "Current Workarounds" section.

### Non-Goals

- Introduce capability options that are not utilized by Gardener in the future.
- Introduce feature flag to control the usage of a capability, e.g. secure boot.

## Proposal

Introduce a top level capabilities map in the CloudProfile `spec.capabilities`.

Capabilities are very specific to the provider and the selected catalog offered by Gardener in the cloud profile.
To minimize complexity and data size in the cloud profile, the capabilities are defined as a map with string keys and string arrays as values.
The key is the capability name and the value is an array of possible values.

```go
type Spec struct {
    Capabilities map[string][]string `json:"capabilities"`
}
```

> [!NOTE] 

> > Due to technical limitations of the protobuf it is not possible to define an array of maps, or a map of string arrays, see [docs](https://protobuf.dev/programming-guides/proto3/#maps). That means ether that the capabilities will be defined internally as a byte array or as a string with comma separated values. To take advantage of most api and code generation features of kubernetes libraries a map of strings seems best, but this consideration is not final and might change in the course of the implementation. I suggest to update the GAP later with whatever proved to be the best solution in terms of performance, maintainability and readability.

Reviewer 2: ```suggestion
> Due to technical limitations of the protobuf, it is not possible to define an array of maps or a map of string arrays, see [docs](https://protobuf.dev/programming-guides/proto3/#maps). That means that the capabilities will either be defined internally as a byte array or as a string with comma-separated values. To take advantage of most API and code generation features of Kubernetes libraries, a map of strings seems best; however, this this consideration is not final and might change in the course of the implementation. I suggest to update the GEP later with whatever proves to be the best solution in terms of performance, maintainability and readability.
```
